### ISSUE TITLE
Negative numerical predictions on always positive values

### ISSUE THREAD

**Issue by atwahsz** [ludwig-ai/ludwig#622]
Negative numerical predictions on always positive values

Is there more parameters that can be added to the output features more than type: numerical . To prevent negative predictions?

**Comment by msaisumanth** [comment]
Right now there's no way to enforce it I think. What would be the way to penalize when the number goes below 0? Can you define an appropriate loss function?

**Comment by w4nderlust** [comment]
@atwahsz if by "on always positive values" you mean "from a model trained on data with output labels containing only positive values and testing it on data from the same distribution as the training one" then that's surprising. I would expect the model to predict values within the range of the ones shown during training.

What I would do is to double check the data first, to make sure there is nothing funny going on there. If everything looks good in the data, then I would provide a minimally reproducible example where that happens (both model + small data, if your data is proprietary please use the dataset synthesizer in `data/dataset_synthesizer.py/` to create one). In that case we can take a look at it.

Finally, to answer your question, at the moment there's no mechanism to enforce the values to be positive, but there is a clipping mechanism so that you can decide the range the values will be clipped to. It is explained in the[ User Guide](https://uber.github.io/ludwig/user_guide/#numerical-output-features-and-decoders). if you look at your data and extract min and max from it you use them as min and max values for clipping. That said, I pretty much believe the error is in the data instead, the clipping should not be needed.

**Comment by atwahsz** [comment]
I added clip: [x,y]

The model trains only improved in one epoch . And when checking predictions it predicts nothing only outputs the maximum value y

**Comment by carlogrisetti** [comment]
That's because, no matter which adjustment to the weights it makes during the training, it will always be clipped to the maximum value of the clip range and thus not improving the prediction result in the validation evaluation.
Clipping only forces the output to be restrained to the min-max range, it does not force the model to "understand" that that's the min-max expected range for that specific domain.
It's basically a shortcut to avoid doing a manual "if x<min then x=min; if x>max then x=max" in the code calling Ludwig (and is thus useful when doing predictions using the CLI, that generates already-clipped values within the various output files)

On a well trained network, using clipping, you can ensure that values that are near the maximum domain accepted solutions don't go over that boundary (ie: if max is 300 and you expect an output of 299, you ensure that it will never go over due to a "misprediction"), but nothing more. You should really doublecheck the training values, and try to test the model feeding a prediction request using one of the same values used in the training. The output should be really close to the training sample at least. If not, you have some issues within the sample data (or the network hasn't been fit enough). If values for the training data are spot-on, it might have overfitted, and try to run a training with a much smaller network and\or consider implementing dropout on the input features.

As a workaround to this behavior, you can try playing with initialization values to try and have the first predictions already within the boundaries you set with the clipping (hence, if you expect a 150 value and the first prediction of the model is 200 within a [0, 300] clipping, the next iteration would be a good one if it goes towards 190 and a bad one if it goes towards 210, "bootstrapping" the learning gradient descent).
One more thing you can do is run the training without the clipping and then "patching" it in the model definition, since it's purely a "support" function implemented by Ludwig and not something that lives within the model itself.

@w4nderlust \ @msaisumanth : an improvement could be to evaluate the model performance *before* clipping the values, but since it's a specific implementation for the numerical feature class, this would mean introducing some sort of "postprocess" stub in the base feature, that will be only extended by the numerical feature containing this clipping operation.

**Comment by w4nderlust** [comment]
I second @carlogrisetti , there's probably something fishy in the data. And for us the only way to help you is to have more details (data to reproduce the issue and the model definition).

Regarding the clipping as postprocessing, I would prefer not to have it in Ludwig and let people do it manually on the outputs, as it seems like an application specific logic, but if there are good arguments i can change my mind.

**Comment by w4nderlust** [comment]
1. yes, you can use the `clip` parameter in the numerical output feature to clip outputs to a specific range.
2. sure, you can specify all the supported losses, and if the loss you want is not available, you can add it.
3. that wouldn-t help, if you have positive inputs and 2 layers with negative weights, you'll get positive outputs, so it's not about the values of the weights.
4. I have no way to know without looking at the data, the model definition and the learning curves. And even in that case, I could only guess :) Anyway, you can tune hyperparameters by modifying your model definition. If you want to do in an automatic way, we are adding an automatic hyperparameter optimization functionality, check the hyperopt branch.

**Comment by w4nderlust** [comment]
@ robertburbidgedata  could you be so kind to retry using the latest code on master? It should have solved the issue.
